Refactor foreach binary ops tests with scalars to use OpInfo

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #52449 [WIP] Add type promotion to a fast path for foreach APIs
* #52448 [WIP] Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* **#51058 Refactor foreach binary ops tests with scalars to use OpInfo**

-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

Commit:
refactor foreach binary ops tests with scalars to use OpInfo

[ghstack-poisoned]

Commit:
Update on "refactor foreach binary ops tests with scalars to use OpInfo"




[ghstack-poisoned]

Commit:
Update on "refactor foreach binary ops tests with scalars to use OpInfo"




[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby committed:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

zou3519 commented on test/test_foreach.py:
```diff
@@ -148,6 +148,129 @@ def test_unary(self, device, dtype, op):
+
+                self.assertEqual(foreach_inplace_exeption, torch_inplace_exeption)
+
+    # Test foreach binary ops with a scalar list
```
nit: let's clarify that this is testing `binary_op(TensorList, List[Scalar])`

zou3519 reviewed (commented):
Logic looks correct. I had two comments:
1. I'm not sure what is happening with the float16 testing (added an in-line comment)
2. In the old code, we had some checks for various error messages that would be raised. Those checks don't exist anymore in the OpInfo tests. Do you think it would be worth it to add some new, non-OpInfo test that tests for the various error message? I am thinking of something like:
```
def test_error_messages(self, device):
    # test subtraction with booleans raises a nice error message

    # test addition with LongTensors but with floating-point alpha raises a nice error message
    
    # ...
```
There are probably better ways to organize this. I'm OK if you want to handle adding tests for the error messages in a follow-up later to unblock progress here

Commit:
Update on "Refactor foreach binary ops tests with scalars to use OpInfo"


-----
Updated foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26103905](https://our.internmc.facebook.com/intern/diff/D26103905)

[ghstack-poisoned]

izdeby commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1029,6 +1029,30 @@ def __init__(self,
+    """Early version of a specialized OpInfo for foreach binary functions"""
+    def __init__(self,
+                 name,
+                 dtypes=all_types_and(torch.bool, torch.half, torch.bfloat16),
```
Yeah, i agree, Its super confusing

mruberry replied:
It is incredibly confusing, sorry. It corresponds to the internal dispatch macro 1:1, though.

zou3519 commented on test/test_foreach.py:
```diff
@@ -148,6 +148,129 @@ def test_unary(self, device, dtype, op):
+                    inplace_name = op.name + "_"
+                    torch_inplace = getattr(torch.Tensor, inplace_name, None)
+
+                    for t in tensors:
+                        torch_inplace(t, scalar)
```
No need to add, just thinking out loud

zou3519 commented on test/test_foreach.py:
```diff
@@ -148,6 +148,129 @@ def test_unary(self, device, dtype, op):
+    def test_binary_ops_scalar(self, device, dtype, op):
+        scalars = [2, 2.2, True, 3 + 5j]
+
+        # Mimics cuda kernel dtype flow.  With fp16/bf16 input, runs in fp32 and casts output back to fp16/bf16.
+        dtype = torch.float32 if (self.device_type == 'cuda' and
+                                  (dtype is torch.float16 or dtype is torch.bfloat16)) else dtype
```
Hmm I see, thanks for the references. Wouldn't it be better to
1. Run the foreach op in with inputs of the original dtype (float16)
2. Convert inputs to the updated dtype (float32), run the reference (PyTorch op in a for-loop), then convert outputs back to float 16
3. Compare the result of 1 and 2

This way we can still test that the foreach APIs accept float16 and bfloat16 tensors without erroring out.
